planetarium/views.py

```diff
@@ -0,0 +1,243 @@
+
+        if name:
+            queryset = queryset.filter(name__icontains=name)
+
+        return queryset
+
+    @extend_schema(
+        parameters=[
+            OpenApiParameter(
+                "name",
+                type=str,
+                description="Filter by name (ex. ?name=Cosmic)",
+                required=False,
+            ),
+        ]
+    )
+    def list(self, request, *args, **kwargs):
+        return super().list(request, *args, **kwargs)
+
+
+class ShowSessionViewSet(viewsets.ModelViewSet):
+    queryset = ShowSession.objects.all()
+    serializer_class = ShowSessionSerializer
+    pagination_class = PageFivePagination
+    permission_classes = (IsAdminOrIfAuthenticatedReadOnly,)
+
+    def get_serializer_class(self):
+        if self.action == "list":
+            return ShowSessionListSerializer
+
+        if self.action == "retrieve":
+            return ShowSessionDetailSerializer
+
+        return self.serializer_class
+
+    def get_queryset(self):
+        queryset = self.queryset.select_related(
+            "astronomy_show", "planetarium_dome"
+        )
+
+        if self.action == "list":
+            queryset = queryset.annotate(
+                tickets_available=(
+                    F("planetarium_dome__seats_in_row")
+                    * F("planetarium_dome__rows")
+                    - Count("tickets")
+                )
+            )
+
+        date = self.request.query_params.get("date")
+        show_title = self.request.query_params.get("show")
+
+        if date:
+            date = datetime.strptime(date, "%Y-%m-%d").date()
+            queryset = queryset.filter(show_time__date=date)
+
+        if show_title:
+            queryset = queryset.filter(
+                astronomy_show__title__icontains=show_title
+            )
+
+        return queryset.distinct()
+
+    @extend_schema(
+        parameters=[
+            OpenApiParameter(
+                "date",
+                type=datetime,
+                description="Filter by date (ex. ?date=2023-10-21)",
+                required=False,
+            ),
+            OpenApiParameter(
+                "show",
+                type={"type": "number"},
+                description="Filter by show title (ex. ?show=Black)",
+            ),
+        ]
+    )
+    def list(self, request, *args, **kwargs):
+        return super().list(request, *args, **kwargs)
+
+
+class ReservationViewSet(
+    mixins.ListModelMixin,
+    mixins.CreateModelMixin,
+    GenericViewSet,
+):
+    queryset = Reservation.objects.all()
+    serializer_class = ReservationSerializer
+    pagination_class = PageFivePagination
+    permission_classes = (IsAuthenticated,)
+
+    def get_queryset(self):
+        queryset = self.queryset.prefetch_related(
+            "tickets__show_session__astronomy_show",
+            "tickets__show_session__planetarium_dome",
+        )
+
+        return queryset.filter(user=self.request.user)
```

Comment:
you can make it in one line, plus is it possible to filter first to make request faster?